gh-140677 Add heatmap visualization to Tachyon sampling profiler

pablogsal opened the pull request:
Introduce a new --heatmap output format that provides line-by-line
execution visualization. The heatmap shows:

- Color-coded execution intensity for each line (cold → warm → hot → very hot)
- Inline sample counts and percentages per line
- Per-file statistics (total samples, hottest line)
- Interactive call graph navigation with caller/callee buttons
- Module type badges (stdlib, site-packages, project code)

Unlike flamegraphs which show call stacks and time distribution, heatmaps
excel at identifying hot code paths within files, understanding line-level
execution patterns, and navigating through call relationships.


<!-- gh-issue-number: gh-140677 -->
* Issue: gh-140677
<!-- /gh-issue-number -->

pablogsal committed:
Add heatmap visualization to Tachyon sampling profiler

Introduce a new --heatmap output format that provides line-by-line
execution visualization. The heatmap shows:

- Color-coded execution intensity for each line (cold → warm → hot → very hot)
- Inline sample counts and percentages per line
- Per-file statistics (total samples, hottest line)
- Interactive call graph navigation with caller/callee buttons
- Module type badges (stdlib, site-packages, project code)

Unlike flamegraphs which show call stacks and time distribution, heatmaps
excel at identifying hot code paths within files, understanding line-level
execution patterns, and navigating through call relationships.

savannahostrowski commented on Lib/profiling/sampling/heatmap_index_template.html:
```diff
@@ -0,0 +1,81 @@
+                </div>
+                <div class="stat-card">
+                    <span class="stat-value"><!-- TOTAL_SAMPLES --></span>
+                    <span class="stat-label">Total Samples</span>
```
We're using "samples" twice in the same view here - once for total samples (i.e., number of snapshots taken during profiling) and the other "samples" to mean the number of lines recorded in that file.

This can be a bit confusing, since file samples can exceed the total number of samples. One idea to help disambiguate might be to add a tooltip or subtext on file samples to say something like "Line samples captured (sum across call stacks)", or rename one of them?

savannahostrowski commented on Lib/profiling/sampling/heatmap_index_template.html:
```diff
@@ -0,0 +1,81 @@
+                            <th onclick="sortTable(3)" style="text-align: right;">Samples ⇅</th>
+                            <th onclick="sortTable(4)" style="text-align: right;">Lines Hit ⇅</th>
+                            <th onclick="sortTable(5)" style="text-align: right;">% of Total ⇅</th>
+                            <th style="text-align: left;">Intensity</th>
```
I don't know if I was sleepy or what but I genuinely had to use brainpower ™️ to understand these columns.

IIUC, these two columns show the same data in different formats. It was not initially obvious to me that % of total wasn't intended to sum to 100% and that the % and bar are actually relative to the hottest file (which should always equal 100%). I feel like my brain expected the bars below the top file to be a proportion of a whole.

A thought here might be to merge the columns into a single column called "Intensity" and then show the percentage number with a coloured background using the same blue --> red heatmap. We could also add a note in the header that this is really the % relative to the hottest file.

savannahostrowski commented on Lib/profiling/sampling/stack_collector.py:
```diff
@@ -287,3 +396,374 @@ def _create_flamegraph_html(self, data):
+            module_name = html.escape(stat['module_name'])
+            module_type = stat['module_type']
+            badge_class = f"badge-{module_type}"
+            bar_width = min(100, stat['percentage'] * 2)  # Scale for visibility
```
I think this calculation isn't quite right? It looks like a file with 46.15% intensity is actually like 92% full because of the `* 2`


<img width="435" height="708" alt="Image" src="https://github.com/user-attachments/assets/25a66ee9-74cc-420d-afaf-a3a8734ade0f" />

savannahostrowski commented on Lib/profiling/sampling/sample.py:
```diff
@@ -632,6 +636,9 @@ def sample(
             filename = filename or f"flamegraph.{pid}.html"
+        case "heatmap":
+            collector = HeatmapCollector(skip_idle=skip_idle)
+            filename = filename or f"heatmap_{pid}"
```
I wonder if folks will want to have this saved by timestamp to compare historical runs?

pablogsal replied:
For now I am keeping the naming consistent with the other modes in the profiler. We can explore this in a separate issue

savannahostrowski reviewed (commented):
Okay, this is _so_ cool! I left a couple of comments, but also wanted to mention that I really loved that you added a minimap to show hotspots and the permalinking on line click; so nice for jumping around and sharing results.

pablogsal committed:
fixup! Add heatmap visualization to Tachyon sampling profiler

pablogsal committed:
Add scroll to caller/callee menu

pablogsal committed:
Merge upstream/main into heatmap branch

pablogsal committed:
fixup! Merge upstream/main into heatmap branch

pablogsal committed:
UX improvements

pablogsal committed:
Imrpovements

pablogsal committed:
MOAR refactor

pablogsal committed:
fixup! MOAR refactor

pablogsal committed:
fixup! fixup! MOAR refactor

pablogsal committed:
Merge remote-tracking branch 'upstream/main' into heatmap

pablogsal committed:
use toggles

pablogsal committed:
Respect the self-time toggle in hot only

pablogsal committed:
Merge remote-tracking branch 'upstream/main' into heatmap

pablogsal committed:
gh-138122: Small fixes to the new tachyon UI

pablogsal committed:
Merge remote-tracking branch 'upstream/main' into heatmap

ivonastojanovic commented on Lib/profiling/sampling/heatmap_collector.py:
```diff
@@ -0,0 +1,1011 @@
+            tree = trees[module_type]
+
+            # Project starts expanded, others start collapsed
+            is_collapsed = module_type in ('stdlib', 'site-packages')
```
Should the _other_ files be collapsed by default? Since this category typically includes special frames or uncategorized files, users are usually more interested in their own project code first. Collapsing these by default might make navigation clearer?

pablogsal committed:
Restyle heatmap like flamegraph

ivonastojanovic commented on Lib/profiling/sampling/heatmap_collector.py:
```diff
@@ -0,0 +1,1038 @@
+
+            self_display = f"{self_samples:,}" if self_samples > 0 else ""
+            cumulative_display = f"{cumulative_samples:,}"
+            tooltip = f"Self: {self_samples:,}, Cumulative: {cumulative_samples:,}"
```
Should we call this _Total_ to stay consistent with the table, since cumulative values are displayed as _Total_ there
```suggestion
            tooltip = f"Self: {self_samples:,}, Total: {cumulative_samples:,}"
```

pablogsal committed:
Refactor css into common files

ivonastojanovic reviewed (commented):
This looks great! I left a few comments, and I’ll experiment with the CSS a bit

ivonastojanovic committed:
Update UI colors and fix intensity legend

Small changes:
- Change navigation button colors for dark theme
- Update toggle button colors and active-state text
- Add missing "→" in the Intensity legend

pablogsal committed:
Fix minimap with hot only toggle

pablogsal committed:
Update Lib/profiling/sampling/heatmap_collector.py

Co-authored-by: ivonastojanovic <80911834+ivonastojanovic@users.noreply.github.com>

pablogsal committed:
Update makefile

pablogsal committed:
Collapse 'other'

ivonastojanovic reviewed (approved):
LGTM 🚀
